MacOS: Add a Homebrew OpenSSL directory configuration option.

junaruga opened the pull request:
This PR is to add Homebrew OpenSSL directory configuration option in MacOS environment. This fixes #1300. It is related to and #1297, as this PR passes CI MacOS cases.

There are 3 commits. The first 2 commits come from another PR:  https://github.com/brianmario/mysql2/pull/1299. The third commit is the main one of this PR.

What do you think? If the PR will be merged, then we need to add this option to the README.md - Configuration options section, and I am happy to help. Note I don't have MacOS environment. I appreciate if someone can help to test this PR on MacOS environment.

I confirmed the CI passed on my forked repository. ([CI link](https://github.com/junaruga/mysql2/actions/runs/3912453010/jobs/6687116518)), and the error case by directory existing check ([CI link](https://github.com/junaruga/mysql2/actions/runs/3911886706/jobs/6685806468#step:7:17)).

In the GitHub Actions build.yml, the way of how to set the matrix environment variables is used in ruby/ruby. You can see [this](https://github.com/ruby/ruby/blob/fc7f8520337be9513d984f67c9ef98ef85d03d0f/.github/workflows/compilers.yml#L220).

## Commit message

MacOS: Add a Homebrew OpenSSL directory configuration option.

This commit is to enable users to set a custom Homebrew openssl directory
configuration option to set the openssl library path on MacOS.

Previously we got the openssl library path by `brew --prefix openssl` on MacOS.
However, there is a case where the directory actually doesn't exist. And this
option enables users to build mysql2 in the case.

```
+ brew --prefix openssl
/usr/local/opt/openssl@3

+ ls -ld /usr/local/opt/openssl*
lrwxr-xr-x  1 runner  admin  36 Dec 16 01:28 /usr/local/opt/openssl -> /usr/local/Cellar/openssl@1.1/1.1.1s
lrwxr-xr-x  1 runner  admin  28 Dec 16 01:19 /usr/local/opt/openssl@1.1 -> ../Cellar/openssl@1.1/1.1.1s
/usr/local/opt/openssl@1.1
```

== How to use ==

In the development, you can run like this.

```
$ RUBY_MYSQL2_BREW_SSL_DIR=/usr/local/opt/openssl@1.1 rake
```

In the installation, you can run like this.

```
$ gem install mysql2 -- --with-brew-ssl-dir=/usr/local/opt/openssl@1.1
```

== Note ==

In this commit, the directory existing check is added in the `extconf.rb`. If
the openssl directory doesn't exist, the build fails with the error message
below.

```
Cannot find library dir(s) /usr/local/opt/openssl@3/lib
```

psulightning commented:
Am I missing it where the builds do not factor in the case without the ENV var set, nor the case where openssl@3 is installed under the Mac OS?

junaruga commented:
> Am I missing it where the builds do not factor in the case without the ENV var set, nor the case where openssl@3 is installed under the Mac OS?

Good question! As far as I know, we don't find how to install openssl3 on MacOS environment because of this issue: https://github.com/brianmario/mysql2/issues/1297.

If the env is not set, and the openssl@3 directory by `brew --prefix openssl` doesn't exist, the mysql2 build fails with the error message "Cannot find library dir(s) /usr/local/opt/openssl@3/lib". I tested it on my forked GitHub Actions.

psulightning commented:
That would be an interesting case also to test against: in essence, the wholly negative case where there is no ENV var, not command line switch and `brew install openssl` never run.

junaruga commented:
> That would be an interesting case also to test against: in essence, the wholly negative case where there is no ENV var, no command line switch, and brew install openssl never run.

If there is no ENV var, there is no command line switch. That is the MacOS cases on the current master branch.

https://github.com/junaruga/mysql2/actions/runs/3912863766/jobs/6688047019
https://github.com/junaruga/mysql2/commit/3ff81996486c054bcf68a3bda70d4c2fe37eb414

sodabrew commented:
Would it be more general to solve this by adding a `--with-ssl-dir`, so the brew command can be passed in like `--with-ssl-dir=$(brew --prefix openssl@1.1` ?

junaruga commented:
Sure. Good point! I will try it. You meant the `--with-ssl-dir` option is implemented in both Mac and Linux cases or only in the Mac cases in this PR?

junaruga commented:
I just rebased this PR with the way you mentioned on the lastet master branch, and updated the commit message too.

The `/usr/local/opt/openssl@1.1` in the `RUBY_MYSQL2_SSL_DIR: '/usr/local/opt/openssl@1.1` in the `.github/workflows/build.yml` is equivalent with the `brew --prefix openssl@1.1`. I changed the logic to set the `RUBY_MYSQL2_SSL_DIR` in the CI yaml. I think this way is more aligned with the current style.

sodabrew commented on ext/mysql2/extconf.rb:
```diff
@@ -28,8 +28,16 @@ def add_ssl_defines(header)
 
 # Homebrew openssl
 if RUBY_PLATFORM =~ /darwin/ && system("command -v brew")
```
Move dir_config(`openssl`) outside of the Darwin block

sodabrew commented on ext/mysql2/extconf.rb:
```diff
@@ -28,8 +28,16 @@ def add_ssl_defines(header)
 if RUBY_PLATFORM =~ /darwin/ && system("command -v brew")
-  openssl_location = `brew --prefix openssl`.strip
-  $LDFLAGS << " -L#{openssl_location}/lib" if openssl_location
+  _, lib = dir_config('ssl')
```
Oh maybe the include dir isn't required -- I don't think any openssl functions are called from code, rather, the libmysqlclient needs to be able to load an openssl library at runtime.

sodabrew commented:
Superseded by #1303
